use-shopping-cart/core/store.js

```diff
@@ -0,0 +1,8 @@
+import { configureStore } from '@reduxjs/toolkit'
+import cartReducer from '../core/slices/cartSlice'
+
+export default configureStore({
+  reducer: {
+    cart: cartReducer
+  }
+})
```

Comment:
We probably won't want to configure the store until we have everything from the developer (i.e. Stripe object, operation mode, success url, failure url, etc).

Reply:
I'm trying to figure out what that would look like. Do you give them a custom store creator that requests these kinda things?

Reply:
So this would all be handled by the inital state from the createCartSlice function, yeah? Just wanted to confirm

Reply:
Actually it would probably just be easier to do it all at the `configureStore` level. What I had in mind was to make the external API for the react portion the same while still using the `core`, any way you can do that works. This is what I came up with:

```js
const initialCartState = {
  mode: "checkout-session",
  lastClicked: "",
  shouldDisplayCart: false,
  stripe: null,
  currency: "USD",
  language: "en-US",
  cartCount: 0,
  totalPrice: 0,
  cartDetails: {}
};
const cartSlice = createSlice({
  name: "cart",
  initialState: initialCartState,
  reducers: {
    clearCart: (state) => {
      state.cartCount = 0;
      state.totalPrice = 0;
      state.cartDetails = {};
    }
  }
});

const {
  reducer,
  actions: { clearCart }
} = cartSlice;

function createShoppingCart(options) {
  return configureStore({
    reducer,
    preloadedState: { ...initialCartState, ...options } // preloadedState is just like initialState
  });
}

function CartProvider({ children, ...props }) {
  // Create our Redux Store just once
  const shoppingCartStore = React.useMemo(
    () => createShoppingCart(props),
    [props]
  );

  // kind of like <Context.Provider> before
  return <Provider store={shoppingCartStore}>{children}</Provider>;
}

// selector is required so, if we want to bypass it, we can provide a default value.
// equalityFn is not required.
function useShoppingCart<mode>(selector = state => state, equalityFn) {
  return useSelector(selector, equalityFn); // kind of like useContext before
}
```

I have a little TypeScript playground of this (it isn't fully implemented obviously because it's just to get some type definitions in the works) if you want to check it out: https://codesandbox.io/s/reduxjstoolkit-usc-typescript-example-e3t99?file=/src/index.tsx

Reply:
@ChrisBrownie55 something seems to be wrong with the playground, but what you have above is pretty cool! I was wondering if memoization would be something that can be handled on the redux side, or if you'd want to do that separately in the React portion for other benefits?

Reply:
@dayhaysoos, The playground shouldn't have much wrong with it. It doesn't render anything though, it's just the code and some types (the types can get messy so they could be doing something weird idk).

I don't believe that the memoization can be done anywhere else. The cart's state comes from `configureStore()` and, by transitivity, `cartSlice.reducer` and the actions come from `cartSlice.actions`. So, we can't create the reducer and add action dispatchors to it at the same time easily. But also, I'm sure each front-end framework has its own way of using Redux and I'd have no clue how memoizing at the store level would affect the use of use-shopping-cart's redux store in those frameworks.